// @sap-ux/odata-service-writer/dist/i18n.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.t = exports.initI18n = void 0;
const i18next_1 = __importDefault(require("i18next"));
const odata_service_writer_i18n_json_1 = __importDefault(require("./translations/odata-service-writer.i18n.json"));
const NS = 'odata-service-writer';
/**
 * Initialize i18next with the translations for this module.
 */
function initI18n() {
    return __awaiter(this, void 0, void 0, function* () {
        yield i18next_1.default.init({
            resources: {
                en: {
                    [NS]: odata_service_writer_i18n_json_1.default
                }
            },
            lng: 'en',
            fallbackLng: 'en',
            defaultNS: NS,
            ns: [NS]
        });
    });
}
exports.initI18n = initI18n;
/**
 * Helper function facading the call call to i18next.
 *
 * @param key i18n key
 * @param options additional options
 * @returns {string} localized string stored for the given key
 */
function t(key, options) {
    return i18next_1.default.t(key, options);
}
exports.t = t;
initI18n();
