// @sap/artifact-management-mdkplugin/dist/src/readers/MdkImageReader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MDK_IMG_PATTERNS = void 0;
const artifact_management_base_1 = require("@sap/artifact-management-base");
const MdkItemReader_1 = __importDefault(require("./MdkItemReader"));
exports.MDK_IMG_PATTERNS = ['Images/*.+(jpg|png|pdf)'];
class MdkImageReader extends MdkItemReader_1.default {
    constructor() {
        super(...arguments);
        this.defaultType = artifact_management_base_1.ItemType.MDKImage;
        this.types = [this.defaultType];
        this.tags = [artifact_management_base_1.Tag.Item, artifact_management_base_1.Tag.Image, artifact_management_base_1.Tag.MDK];
        this.matchConditions = { requiredFilePatterns: exports.MDK_IMG_PATTERNS };
    }
    readEntities(paths) {
        return Promise.resolve(paths.map(() => null));
    }
}
exports.default = MdkImageReader;
