// @sap/artifact-management-mdkplugin/dist/src/readers/MdkRuleReader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MDK_RULE_PATTERNS = void 0;
const artifact_management_base_1 = require("@sap/artifact-management-base");
const MdkItemReader_1 = __importDefault(require("./MdkItemReader"));
exports.MDK_RULE_PATTERNS = ['Rules/**/*.js'];
class MdkRuleReader extends MdkItemReader_1.default {
    constructor() {
        super(...arguments);
        this.defaultType = artifact_management_base_1.ItemType.MDKRule;
        this.types = [this.defaultType];
        this.tags = [artifact_management_base_1.Tag.Item, artifact_management_base_1.Tag.Rule, artifact_management_base_1.Tag.MDK];
        this.matchConditions = { requiredFilePatterns: exports.MDK_RULE_PATTERNS };
    }
    readEntities(paths, fs) {
        return Promise.all(paths.map(p => fs.readTextFile(p)));
    }
}
exports.default = MdkRuleReader;
