// @sap/artifact-management-mdkplugin/dist/src/readers/MdkStyleReader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MDK_STYLE_PATTERNS = void 0;
const artifact_management_base_1 = require("@sap/artifact-management-base");
const MdkItemReader_1 = __importDefault(require("./MdkItemReader"));
exports.MDK_STYLE_PATTERNS = ['Styles/*.+(less|css|nss|json)'];
class MdkStyleReader extends MdkItemReader_1.default {
    constructor() {
        super(...arguments);
        this.defaultType = artifact_management_base_1.ItemType.MDKStyle;
        this.types = [this.defaultType];
        this.tags = [artifact_management_base_1.Tag.Item, artifact_management_base_1.Tag.Style, artifact_management_base_1.Tag.MDK];
        this.patterns = exports.MDK_STYLE_PATTERNS;
        this.matchConditions = { requiredFilePatterns: exports.MDK_STYLE_PATTERNS };
    }
    readEntities(paths) {
        return Promise.resolve(paths.map(() => null));
    }
}
exports.default = MdkStyleReader;
