// @sap/artifact-management-base/dist/src/commons/PropertyFile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class PropertyFile {
    constructor(fs, path) {
        this.properties = {};
        this.fs = fs;
        this.path = path;
    }
    async load() {
        this.properties = {};
        const text = await this.fs.readTextFile(this.path);
        for (const line of text.split(/\n\r?|\r\n/)) {
            // Comment
            if (line.match(/^\s*#/))
                continue;
            // Property
            const match = line.match(/^\s*((\\.|[^= \\])*)\s*=\s*(.*)/);
            if (!match)
                continue;
            try {
                this.properties[this.unescape(match[1])] = this.unescape(match[3]);
            }
            catch (error) {
                // broken unicode
                continue;
            }
        }
    }
    unescape(str) {
        return str
            .replace(/\\(u([0-9a-f]+)|[^u])/ig, (str, value) => {
            if (value.startsWith('u'))
                return JSON.parse(`"\\${value}"`);
            else
                return value;
        });
    }
}
exports.default = PropertyFile;
