// @sap/artifact-management-base/dist/src/cp/UI5ManifestHelper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const PropertyFile_1 = __importDefault(require("../commons/PropertyFile"));
const path_1 = __importDefault(require("path"));
class UI5ManifestHelper {
    constructor(manifest) {
        this.manifest = manifest;
    }
    async loadi18nProperties(manifest) {
        const propertiesFilePath = manifest.data['sap.app'].i18n;
        if (propertiesFilePath) {
            this.propertiesFile = new PropertyFile_1.default(manifest.fs, path_1.default.join('webapp', propertiesFilePath));
            await this.propertiesFile.load();
        }
    }
    getValue(str, properties) {
        if (str === '') {
            return '<missing text>';
        }
        return str.replace(/\{\{([^\}]+)\}\}/g, (match, name) => {
            if (name in properties) {
                return properties[name];
            }
            else {
                return `<missing text "${name}">`;
            }
        });
    }
    async getLocalisedValue(property) {
        var _a;
        if (this.propertiesFile === undefined) {
            await this.loadi18nProperties(this.manifest);
        }
        const properties = ((_a = this.propertiesFile) === null || _a === void 0 ? void 0 : _a.properties) || {};
        return this.getValue(property, properties);
    }
}
exports.default = UI5ManifestHelper;
