// @sap/artifact-management-base/dist/src/cp/UI5RunUI.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const UIManifest_1 = __importDefault(require("./UIManifest"));
const UI5ManifestHelper_1 = __importDefault(require("./UI5ManifestHelper"));
const ModuleType_1 = require("../definitions/ModuleType");
class UI5RunUI {
    getType() {
        return ModuleType_1.ModuleType.UI5;
    }
    async applications({ fs }) {
        const applications = [];
        const manifest = new UIManifest_1.default(fs, 'webapp/manifest.json');
        await manifest.load();
        const title = manifest.data['sap.app'].title || manifest.data['sap.app'].id;
        const description = manifest.data['sap.app'].description;
        const inbounds = manifest.crossNavigationInbounds();
        const inboundKeys = Object.keys(inbounds).sort();
        let inbound;
        if (inboundKeys[0]) {
            inbound = inbounds[inboundKeys[0]];
        }
        else {
            inbound = {
                semanticObject: this.makeValidSemanticObject(manifest.data['sap.app'].id),
                action: 'open',
            };
        }
        const ui5manifestHelper = new UI5ManifestHelper_1.default(manifest);
        applications.push({
            id: manifest.data['sap.app'].id,
            title: await ui5manifestHelper.getLocalisedValue(title),
            description: description ? await ui5manifestHelper.getLocalisedValue(description) : '',
            ui5Component: manifest.data['sap.app'].id,
            path: fs.relativeToRoot('webapp'),
            action: inbound.action,
            semanticObject: inbound.semanticObject,
            deploymentId: `${manifest.data['sap.app'].id.replace(/[-.]/g, '')}-${manifest.data['sap.app'].applicationVersion.version}`,
        });
        return applications;
    }
    makeValidSemanticObject(id) {
        return id.replace(/[^a-zA-Z0-9]+/g, (str) => `_${Buffer.from(str).toString('hex')}_`);
    }
}
exports.default = UI5RunUI;
