// @sap/artifact-management-mdkplugin/dist/src/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InfoList = exports.MdkRunUI = exports.MdkServiceBuild = exports.MdkModuleGenerator = exports.MdkModuleReader = void 0;
const MdkModuleGenerator_1 = __importDefault(require("./generators/MdkModuleGenerator"));
const MdkModuleReader_1 = __importDefault(require("./readers/MdkModuleReader"));
const MdkRunUI_1 = __importDefault(require("./MdkRunUI"));
const MdkServiceBuild_1 = __importDefault(require("./MdkServiceBuild"));
const MDKPlugin = {
    name: '@sap/mdk-project-plugin',
    getDependencies() {
        return [];
    },
    // TODO: Need to incorporate RunUI!!
    getProjectReaders() {
        return [];
    },
    getModuleReaders() {
        return [new MdkModuleReader_1.default()];
    },
    getProjectManifestGenerators() {
        return [];
    },
    getModuleManifestGenerators() {
        return [new MdkModuleGenerator_1.default()];
    },
    getModuleRunner() {
        return new MdkRunUI_1.default();
    },
    getModuleBuilder() {
        return new MdkServiceBuild_1.default();
    },
};
exports.default = MDKPlugin;
var MdkModuleReader_2 = require("./readers/MdkModuleReader");
Object.defineProperty(exports, "MdkModuleReader", { enumerable: true, get: function () { return __importDefault(MdkModuleReader_2).default; } });
var MdkModuleGenerator_2 = require("./generators/MdkModuleGenerator");
Object.defineProperty(exports, "MdkModuleGenerator", { enumerable: true, get: function () { return __importDefault(MdkModuleGenerator_2).default; } });
var MdkServiceBuild_2 = require("./MdkServiceBuild");
Object.defineProperty(exports, "MdkServiceBuild", { enumerable: true, get: function () { return __importDefault(MdkServiceBuild_2).default; } });
var MdkRunUI_2 = require("./MdkRunUI");
Object.defineProperty(exports, "MdkRunUI", { enumerable: true, get: function () { return __importDefault(MdkRunUI_2).default; } });
var InfoList_1 = require("./InfoList");
Object.defineProperty(exports, "InfoList", { enumerable: true, get: function () { return InfoList_1.InfoList; } });
