// @sap/artifact-management-mdkplugin/dist/src/InfoList.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InfoList = void 0;
exports.InfoList = {
    ERR_APPLICATION_NAME_UNDEFINED: {
        code: 'mdk_001',
        description: 'Cannot find _Name in Application.app, hence name of this module is undefined',
    },
    ERR_CORRUPT_FILE: {
        code: 'mdk_002',
        description: 'Error occurred while reading this file. File may be corrupted',
    },
    ERR_LCAP_UNDEFINED: {
        code: 'mdk_003',
        description: 'Cannot find LCAP in .project.json, hence sourceTemplate and service links are undefined',
    },
    ERR_SERVICE_URL_DESTINATION_NAME_UNDEFINED: {
        code: 'mdk_004',
        description: 'Cannot find ServiceUrl and DestinationName, hence name and ref of this item are undefined',
    },
    ERR_SOURCETEMPLATE_ID_UNDEFINED: {
        code: 'mdk_005',
        description: 'Cannot find LCAP.sourceTemplate.id in .project.json, hence sourceTemplate id is undefined',
    },
    ERR_CONFIG_CORRUPT_FILE: {
        code: 'mdk_006',
        description: 'Error occurred while reading Web/config.json file. File may be corrupted',
    },
    ERR_CONFIG_MISSING: {
        code: 'mdk_007',
        description: 'Error occurred while reading Web/config.json file. File may be missing',
    },
};
