// @sap/artifact-management/dist/src/plugins/cap/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InfoList = void 0;
const CapModuleReader_1 = __importDefault(require("./readers/CapModuleReader"));
const CapProjectReader_1 = __importDefault(require("./readers/CapProjectReader"));
const CapProjectGenerator_1 = __importDefault(require("./generators/CapProjectGenerator"));
const CapModuleGenerator_1 = __importDefault(require("./generators/CapModuleGenerator"));
const CapRun_1 = __importDefault(require("./CapRun"));
const CapPlugin = {
    name: '@sap/cap-project-plugin',
    getDependencies() {
        return [
            'ui5',
            'mdk',
            'xsuaa',
        ];
    },
    getProjectReaders() {
        return [new CapProjectReader_1.default()];
    },
    getModuleReaders() {
        return [new CapModuleReader_1.default()];
    },
    getProjectManifestGenerators() {
        return [new CapProjectGenerator_1.default()];
    },
    getModuleManifestGenerators() {
        return [new CapModuleGenerator_1.default()];
    },
    getProjectRunner() {
        return new CapRun_1.default();
    },
};
exports.default = CapPlugin;
var InfoList_1 = require("./InfoList");
Object.defineProperty(exports, "InfoList", { enumerable: true, get: function () { return InfoList_1.InfoList; } });
