// @sap/artifact-management-base/dist/src/project/ItemLink.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isLinkTypeReverseValid = exports.LinkTypeReverse = exports.LinkType = void 0;
var LinkType;
(function (LinkType) {
    LinkType["DEPENDS_ON"] = "DEPENDS_ON";
    LinkType["HAS_DEPENDENCY"] = "HAS_DEPENDENCY";
    LinkType["TRANSITIVELY_DEPENDS_ON"] = "TRANSITIVELY_DEPENDS_ON";
    LinkType["HAS_TRANSITIVE_DEPENDENCY"] = "HAS_TRANSITIVE_DEPENDENCY";
    LinkType["PARENT_OF"] = "PARENT_OF";
    LinkType["CHILD_OF"] = "CHILD_OF";
    LinkType["ASSOCIATES_TO_ONE"] = "ASSOCIATES_TO_ONE";
    LinkType["ASSOCIATES_TO_MANY"] = "ASSOCIATES_TO_MANY";
    LinkType["HAS_ASSOCIATION_WITH_ONE"] = "HAS_ASSOCIATION_WITH_ONE";
    LinkType["HAS_ASSOCIATION_WITH_MANY"] = "HAS_ASSOCIATION_WITH_MANY";
    LinkType["HAS_COMPOSITION_WITH_ONE"] = "HAS_COMPOSITION_WITH_ONE";
    LinkType["HAS_COMPOSITION_WITH_MANY"] = "HAS_COMPOSITION_WITH_MANY";
    LinkType["SELF"] = "SELF";
})(LinkType = exports.LinkType || (exports.LinkType = {}));
exports.LinkTypeReverse = {
    [LinkType.CHILD_OF]: LinkType.PARENT_OF,
    [LinkType.PARENT_OF]: LinkType.CHILD_OF,
    [LinkType.DEPENDS_ON]: LinkType.HAS_DEPENDENCY,
    [LinkType.HAS_DEPENDENCY]: LinkType.DEPENDS_ON,
    [LinkType.TRANSITIVELY_DEPENDS_ON]: LinkType.HAS_TRANSITIVE_DEPENDENCY,
    [LinkType.HAS_TRANSITIVE_DEPENDENCY]: LinkType.TRANSITIVELY_DEPENDS_ON,
    [LinkType.ASSOCIATES_TO_ONE]: LinkType.SELF,
    [LinkType.ASSOCIATES_TO_MANY]: LinkType.SELF,
    [LinkType.HAS_ASSOCIATION_WITH_ONE]: LinkType.SELF,
    [LinkType.HAS_ASSOCIATION_WITH_MANY]: LinkType.SELF,
    [LinkType.HAS_COMPOSITION_WITH_ONE]: LinkType.SELF,
    [LinkType.HAS_COMPOSITION_WITH_MANY]: LinkType.SELF,
    [LinkType.SELF]: LinkType.SELF,
};
function isLinkTypeReverseValid(linkType) {
    if (exports.LinkTypeReverse[linkType] === LinkType.SELF) {
        return false;
    }
    else {
        return true;
    }
}
exports.isLinkTypeReverseValid = isLinkTypeReverseValid;
