// @sap/artifact-management-base/dist/src/project/ModuleReader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModuleReader = void 0;
/**
 * Read modules and its entities
 *
 * There are two ways to implement a ModuleReader:
 * 1. Read complete module and entities in one call --> re-implement read()
 * 2. Read module header and entities separately --> re-implement readModule()
 *
 * In both cases entityReaders() should return the entity readers with the super set of their handled types and tags and
 * the relevant file pattern to recognize changes. This is relevant to have a fast project update in BAS.
 */
class ModuleReader {
    constructor() {
        this.moduleType = undefined;
        this.matchConditions = {};
        this.itemReaders = [];
        this.tags = [];
    }
    /**
     *
     * @param fs
     *
     * Read module header information (without entities).
     *
     * Re-implement this method if the module header and its entities should be read separately.
     */
    read(options) {
        throw new Error('Method not implemented.');
    }
}
exports.ModuleReader = ModuleReader;
ModuleReader.type = 'sap.project.plugin.type.module-reader';
