// @sap/artifact-management-base/dist/src/project/PluginEntryTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDetectionPriority = exports.DetectionOrder = exports.DetectionMechanism = void 0;
/**
 * Depending on the DetectionMechanism the priority of the plugin execution will be determined.
 */
var DetectionMechanism;
(function (DetectionMechanism) {
    /** Uniquely identified */
    DetectionMechanism["UNIQUE_IDENTIFICATION"] = "UNIQUE_IDENTIFICATION";
    /** Identified by searching for file pattern */
    DetectionMechanism["FILE_PATTERN"] = "FILE_PATTERN";
    /** Fallback, no specific detection */
    DetectionMechanism["FALLBACK"] = "FALLBACK";
    /** Ignored for detection */
    DetectionMechanism["IGNORE"] = "IGNORE";
})(DetectionMechanism = exports.DetectionMechanism || (exports.DetectionMechanism = {}));
exports.DetectionOrder = [
    DetectionMechanism.UNIQUE_IDENTIFICATION,
    DetectionMechanism.FILE_PATTERN,
    DetectionMechanism.FALLBACK,
];
const DetectionPriority = exports.DetectionOrder.reduce((map, mechanism, index) => {
    map[mechanism] = index + 1;
    return map;
}, {});
function getDetectionPriority(mechanism) {
    return DetectionPriority[mechanism] || exports.DetectionOrder.length + 1;
}
exports.getDetectionPriority = getDetectionPriority;
