// @sap/artifact-management-base/dist/src/project/ProjectWrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Tag_1 = require("../definitions/Tag");
const ProjectReader_1 = require("./ProjectReader");
class ProjectWrapper extends ProjectReader_1.ProjectReader {
    constructor(moduleReader) {
        super({
            requiredFilePatterns: moduleReader.matchConditions.requiredFilePatterns,
        });
        this.tags = [Tag_1.Tag.Project, Tag_1.Tag.CAP];
        this.moduleReader = moduleReader;
        this.tags = [Tag_1.Tag.Project, ...moduleReader.tags].filter(val => val !== Tag_1.Tag.Module);
    }
    getType() {
        return this.moduleReader.getType();
    }
    getDetectionMechanism() {
        return this.moduleReader.getDetectionMechanism();
    }
    async read({ fs }) {
        const pattern = this.moduleReader.matchConditions.requiredFilePatterns ? this.moduleReader.matchConditions.requiredFilePatterns : [];
        const moduleData = await this.moduleReader.read({ fs: fs, matchedFiles: { pattern: pattern } });
        if (!moduleData) {
            return;
        }
        const prefix = moduleData.name;
        const name = moduleData.name;
        return {
            type: this.getType(),
            path: fs.basePath,
            prefix,
            cloudService: `${prefix}`,
            name,
            tags: this.tags,
        };
    }
}
exports.default = ProjectWrapper;
