// @sap/artifact-management/dist/src/plugins/cap/generators/CapLaunchpadGenerator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const artifact_management_base_1 = require("@sap/artifact-management-base");
class CapLaunchpadGenerator {
    getType() {
        return artifact_management_base_1.ModuleType.CAP;
    }
    async provideResources(context) {
        return [];
    }
    async beforeBuildCommands() {
        return [];
    }
    async getModuleManifest(context) {
        const zipName = 'launchpad.zip';
        return {
            name: context.nameFactory.moduleName('launchpad'),
            type: artifact_management_base_1.MtaModuleType.HTML5,
            path: context.fs().relativeToRoot('build/launchpad'),
            'build-parameters': {
                builder: 'custom',
                commands: [
                    'npm install',
                    'npx -p @ui5/cli ui5 build --dest dist --include-task=generateManifestBundle --include-task=generateCachebusterInfo',
                    `bash -c "cd dist && npx bestzip ../../../build/dist/${zipName} *"`,
                    'rm -rf dist',
                ],
                _artifactName: zipName,
                'supported-platforms': [],
                'build-result': '../../build/dist',
            },
        };
    }
    requires() {
        return [];
    }
}
exports.default = CapLaunchpadGenerator;
