// @sap/artifact-management/dist/src/plugins/cap/generators/CapUIDeployerGenerator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const artifact_management_base_1 = require("@sap/artifact-management-base");
class CapUIDeployerGenerator {
    getType() {
        return artifact_management_base_1.ModuleType.CAP;
    }
    async provideResources() {
        return [];
    }
    async beforeBuildCommands() {
        return ['rm -rf build/ui-deployer',
            'mkdir -p build/ui-deployer',
            'rm -rf build/dist',
            'mkdir -p build/dist',
        ];
    }
    async getModuleManifest(context) {
        context.moduleName = context.nameFactory.auxModuleName('ui-deployer');
        const templatePath = __dirname + '/';
        const module = await artifact_management_base_1.MtaTemplate.build(templatePath, 'mta-ui-deployer-template.yaml', context);
        const artifacts = [];
        for (const mtaModule of context.builtModules) {
            if (mtaModule.type === 'html5') {
                artifacts.push({
                    name: mtaModule.name,
                    artifacts: [mtaModule['build-parameters']._artifactName],
                    'target-path': 'resources',
                });
            }
        }
        module['build-parameters'].requires = artifacts;
        return module;
    }
    requires() {
        return [{
                serviceName: artifact_management_base_1.ServiceCatalog.html5_apps_repo_host.name,
                servicePlan: artifact_management_base_1.ServiceCatalog.html5_apps_repo_host.plan,
                parameters: { 'content-target': true },
            }];
    }
}
exports.default = CapUIDeployerGenerator;
