// @sap/bas-sdk/src/utils/devspace-utils.ts
/**
 * Exposed devspace information
 */
export interface DevspaceInfo {
  devspaceDisplayName: string;
  devspaceOrigin: string;
  pack: string;
  packDisplayName: string;
}

/**
 * Full interface for results of call ENV_H2O_URL/ws-manager/api/v1/workspace/${workspaceId}
 */
export interface DevspaceApiResults {
  devSpaceOriginLabel?: string;
  config: {
    id: string;
    username: string;
    labels: {
      controllerID: string;
      "workspace.devx.sap.com/persistence-pvc": string;
      "ws-manager.devx.sap.com/deleted": string;
      "ws-manager.devx.sap.com/displayname": string;
      "ws-manager.devx.sap.com/origin": string;
      "ws-manager.devx.sap.com/providerTenantId": string;
      "ws-manager.devx.sap.com/username": string;
      "ws-manager.devx.sap.com/zoneid": string;
      "ws-manager.devx.sap.com/LCAP"?: string;
    };
    annotations: {
      optionalExtensions: string;
      pack: string;
      packTagline: string;
      technicalExtensions: string;
      "workspace.devx.sap.com/ignoreExtensionRemovalOps"?: string;
      "workspace.devx.sap.com/lastOnlineTime": string;
      "workspace.devx.sap.com/ownerFullName": string;
      "workspace.devx.sap.com/plan": string;
      "workspace.devx.sap.com/ws-controller-processed-gen": string;
      "ws-manager.devx.sap.com/startedAt": string;
    };
    creationtimestamp: string;
    encrypted: boolean;
  };
  runtime: {
    status: string;
    url: {
      startup: string;
      theia: string;
    };
    baseUrl: string;
    statusLastUpdateTime: string;
    statusLastUpdateGeneration: number;
    backendReadiness: { name: string; readiness: boolean }[];
  };
  diskusage: {
    used: number;
    free: number;
    iused: number;
    ifree: number;
    timestamp: string;
  };
}

/**
 * Create flat structure of exposed workspace information from given api call results.
 *
 * @param wsSettings - full workspace  workspace
 * @returns - workspace information
 */
export function flattenDevspaceInfo(
  devspaceResults: DevspaceApiResults
): DevspaceInfo {
  return {
    devspaceDisplayName:
      devspaceResults?.config?.labels?.[
        "ws-manager.devx.sap.com/displayname"
      ] || "",
    devspaceOrigin:
      devspaceResults?.devSpaceOriginLabel ||
      devspaceResults?.config?.labels?.["ws-manager.devx.sap.com/LCAP"] ||
      "",
    pack: devspaceResults?.config?.annotations?.pack || "",
    packDisplayName: devspaceResults?.config?.annotations?.packTagline || "",
  };
}
