// @sap-ux/fiori-elements-writer/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generate = void 0;
const path_1 = require("path");
const ejs_1 = require("ejs");
const ui5_application_writer_1 = require("@sap-ux/ui5-application-writer");
const odata_service_writer_1 = require("@sap-ux/odata-service-writer");
const packageConfig_1 = require("./packageConfig");
const cloneDeep_1 = __importDefault(require("lodash/cloneDeep"));
const validate_1 = require("./validate");
const defaults_1 = require("./data/defaults");
const templateAttributes_1 = require("./data/templateAttributes");
const manifestSettings_1 = require("./data/manifestSettings");
const semver_1 = __importDefault(require("semver"));
/**
 * Generate a UI5 application based on the specified Fiori Elements floorplan template.
 *
 * @param basePath - the absolute target path where the application will be generated
 * @param data - configuration to generate the Fiori elements application
 * @param fs - an optional reference to a mem-fs editor
 * @returns Reference to a mem-fs-editor
 */
function generate(basePath, data, fs) {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    return __awaiter(this, void 0, void 0, function* () {
        // Clone rather than modifying callers refs
        const feApp = cloneDeep_1.default(data);
        // Ensure input data contains at least the manadatory properties required for app genertation
        validate_1.validateRequiredProperties(feApp);
        defaults_1.setAppDefaults(feApp);
        fs = yield ui5_application_writer_1.generate(basePath, feApp, fs);
        feApp.template.settings = defaults_1.setDefaultTemplateSettings(feApp.template, feApp.service.version);
        // This is done after `generateUi5Project` since defaults are set if values are not provided
        validate_1.validateApp(feApp);
        yield odata_service_writer_1.generate(basePath, feApp.service, fs);
        const templateOptions = {
            changesPreview: ((_a = feApp.ui5) === null || _a === void 0 ? void 0 : _a.version) ? semver_1.default.lt(semver_1.default.coerce((_b = feApp.ui5) === null || _b === void 0 ? void 0 : _b.version), templateAttributes_1.changesPreviewToVersion)
                : false,
            changesLoader: feApp.service.version === odata_service_writer_1.OdataVersion.v2
        };
        // Add new files from templates e.g.
        const rootTemplatesPath = path_1.join(__dirname, '..', 'templates');
        // Add templates common to all template types
        fs.copyTpl(path_1.join(rootTemplatesPath, 'common', 'add', '**/*.*'), basePath, Object.assign(Object.assign({}, feApp), { templateOptions,
            escapeFLPText: templateAttributes_1.escapeFLPText }));
        // Extend common files
        const packagePath = path_1.join(basePath, 'package.json');
        // Extend package.json
        fs.extendJSON(packagePath, JSON.parse(ejs_1.render(fs.read(path_1.join(rootTemplatesPath, 'common', 'extend', 'package.json')), feApp)));
        const templateVersionPath = path_1.join(rootTemplatesPath, `v${(_c = feApp.service) === null || _c === void 0 ? void 0 : _c.version}`);
        // Copy version specific common templates and version specific, floorplan specific templates
        [path_1.join(templateVersionPath, 'common', 'add'), path_1.join(templateVersionPath, feApp.template.type, 'add')].forEach((templatePath) => {
            fs.copyTpl(path_1.join(templatePath, '**/*.*'), basePath, feApp, {}, { ignoreNoMatch: true });
        });
        // Update manifest.json with template specific settings
        manifestSettings_1.extendManifestJson(fs, basePath, rootTemplatesPath, feApp);
        const packageJson = JSON.parse(fs.read(packagePath));
        packageJson.scripts = Object.assign(packageJson.scripts, Object.assign({}, packageConfig_1.getPackageJsonTasks({
            localOnly: !((_d = feApp.service) === null || _d === void 0 ? void 0 : _d.url),
            addMock: !!((_e = feApp.service) === null || _e === void 0 ? void 0 : _e.metadata),
            sapClient: (_f = feApp.service) === null || _f === void 0 ? void 0 : _f.client,
            flpAppId: feApp.app.flpAppId,
            startFile: (_g = data === null || data === void 0 ? void 0 : data.app) === null || _g === void 0 ? void 0 : _g.startFile,
            localStartFile: (_h = data === null || data === void 0 ? void 0 : data.app) === null || _h === void 0 ? void 0 : _h.localStartFile
        })));
        fs.writeJSON(packagePath, packageJson);
        return fs;
    });
}
exports.generate = generate;
__exportStar(require("./types"), exports);
