// @sap-ux/fiori-elements-writer/dist/validate.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateRequiredProperties = exports.validateApp = void 0;
const semver_1 = __importDefault(require("semver"));
const templateAttributes_1 = require("./data/templateAttributes");
const i18n_1 = require("./i18n");
const types_1 = require("./types");
/**
 * Validates a selection of specified app settings.
 *
 * @param feApp - Fiori elements application configuration
 */
function validateApp(feApp) {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    // Validate odata version
    if (feApp.service.version &&
        !templateAttributes_1.TemplateTypeAttributes[feApp.template.type].supportedODataVersions.includes(feApp.service.version)) {
        throw new types_1.ValidationError(i18n_1.t('error.unsupportedOdataVersion', {
            serviceVersion: feApp.service.version,
            templateType: feApp.template.type
        }));
    }
    // Validate ui5 versions if present (defaults will apply otherwise)
    let ui5Version = null;
    if ((_a = feApp.ui5) === null || _a === void 0 ? void 0 : _a.version) {
        ui5Version = semver_1.default.coerce((_b = feApp.ui5) === null || _b === void 0 ? void 0 : _b.version);
        if (!ui5Version) {
            throw new types_1.ValidationError(i18n_1.t('error.invalidUI5Version', { versionProperty: 'version', ui5Version: (_c = feApp.ui5) === null || _c === void 0 ? void 0 : _c.version }));
        }
    }
    let minUI5Version;
    const minRequiredUI5Version = templateAttributes_1.TemplateTypeAttributes[feApp.template.type].minimumUi5Version[feApp.service.version];
    if ((_d = feApp.ui5) === null || _d === void 0 ? void 0 : _d.minUI5Version) {
        minUI5Version = semver_1.default.coerce((_e = feApp.ui5) === null || _e === void 0 ? void 0 : _e.minUI5Version);
        if (!minUI5Version) {
            throw new types_1.ValidationError(i18n_1.t('error.invalidUI5Version', { versionProperty: 'minUI5Version', ui5Version: (_f = feApp.ui5) === null || _f === void 0 ? void 0 : _f.minUI5Version }));
        }
    }
    else {
        minUI5Version = semver_1.default.coerce(minRequiredUI5Version);
    }
    if (ui5Version && semver_1.default.lt(ui5Version, minRequiredUI5Version)) {
        throw new types_1.ValidationError(i18n_1.t('error.unsupportedUI5Version', {
            versionProperty: 'version',
            ui5Version: (_g = feApp.ui5) === null || _g === void 0 ? void 0 : _g.version,
            templateType: feApp.template.type,
            minRequiredUI5Version
        }));
    }
    if (semver_1.default.lt(minUI5Version, minRequiredUI5Version)) {
        throw new types_1.ValidationError(i18n_1.t('error.unsupportedUI5Version', {
            versionProperty: 'minUI5Version',
            ui5Version: (_h = feApp.ui5) === null || _h === void 0 ? void 0 : _h.minUI5Version,
            templateType: feApp.template.type,
            minRequiredUI5Version
        }));
    }
}
exports.validateApp = validateApp;
/**
 * Validates the specified FioriElementsApp contains the required properties.
 *
 * @param feApp - Fiori elements application configuration
 */
function validateRequiredProperties(feApp) {
    // Service is mandatory for generation of Fiori Elements apps
    if (!feApp.service) {
        throw new types_1.ValidationError(i18n_1.t('error.missingRequiredProperty', {
            propertyName: 'FioriElementsApp.service'
        }));
    }
}
exports.validateRequiredProperties = validateRequiredProperties;
