// @sap/artifact-management-base/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Info = exports.TestUtil = exports.MtaTemplate = exports.UI5RunUI = exports.UI5ManifestHelper = exports.UIManifest = exports.CapApi = exports.DestinationData = exports.Service = exports.NameFactory = exports.RunUI = exports.NodeFileSystem = exports.ProjectFileSystem = exports.Util = exports.PropertyFile = exports.consoleLogger = exports.ProjectAPILogger = exports.NPMUtil = exports.makeType = exports.ItemFilterImpl = exports.EventType = exports.ErrorWithLocation = exports.isLinkTypeReverseValid = exports.LinkType = exports.LinkTypeReverse = exports.getDetectionPriority = exports.DetectionMechanism = exports.FilterMatchCriteria = exports.ProjectWrapper = exports.ProjectRunner = exports.ModuleAutoBuilder = exports.ItemType = exports.PluginNotFoundError = exports.Tag = exports.ModuleType = exports.ProjectType = void 0;
var ProjectType_1 = require("./definitions/ProjectType");
Object.defineProperty(exports, "ProjectType", { enumerable: true, get: function () { return ProjectType_1.ProjectType; } });
var ModuleType_1 = require("./definitions/ModuleType");
Object.defineProperty(exports, "ModuleType", { enumerable: true, get: function () { return ModuleType_1.ModuleType; } });
var Tag_1 = require("./definitions/Tag");
Object.defineProperty(exports, "Tag", { enumerable: true, get: function () { return Tag_1.Tag; } });
var PluginError_1 = require("./definitions/PluginError");
Object.defineProperty(exports, "PluginNotFoundError", { enumerable: true, get: function () { return __importDefault(PluginError_1).default; } });
var ItemType_1 = require("./definitions/ItemType");
Object.defineProperty(exports, "ItemType", { enumerable: true, get: function () { return ItemType_1.ItemType; } });
__exportStar(require("./project/ProjectReader"), exports);
__exportStar(require("./project/ModuleReader"), exports);
__exportStar(require("./project/ProjectReaderHelper"), exports);
__exportStar(require("./project/Plugin"), exports);
__exportStar(require("./project/ModuleGeneratorContext"), exports);
var ModuleAutoBuilder_1 = require("./project/ModuleAutoBuilder");
Object.defineProperty(exports, "ModuleAutoBuilder", { enumerable: true, get: function () { return __importDefault(ModuleAutoBuilder_1).default; } });
var ProjectRunner_1 = require("./project/ProjectRunner");
Object.defineProperty(exports, "ProjectRunner", { enumerable: true, get: function () { return __importDefault(ProjectRunner_1).default; } });
var ProjectWrapper_1 = require("./project/ProjectWrapper");
Object.defineProperty(exports, "ProjectWrapper", { enumerable: true, get: function () { return __importDefault(ProjectWrapper_1).default; } });
var ItemFilter_1 = require("./project/ItemFilter");
Object.defineProperty(exports, "FilterMatchCriteria", { enumerable: true, get: function () { return ItemFilter_1.FilterMatchCriteria; } });
var PluginEntryTypes_1 = require("./project/PluginEntryTypes");
Object.defineProperty(exports, "DetectionMechanism", { enumerable: true, get: function () { return PluginEntryTypes_1.DetectionMechanism; } });
Object.defineProperty(exports, "getDetectionPriority", { enumerable: true, get: function () { return PluginEntryTypes_1.getDetectionPriority; } });
var ItemLink_1 = require("./project/ItemLink");
Object.defineProperty(exports, "LinkTypeReverse", { enumerable: true, get: function () { return ItemLink_1.LinkTypeReverse; } });
Object.defineProperty(exports, "LinkType", { enumerable: true, get: function () { return ItemLink_1.LinkType; } });
Object.defineProperty(exports, "isLinkTypeReverseValid", { enumerable: true, get: function () { return ItemLink_1.isLinkTypeReverseValid; } });
var ErrorWithLocation_1 = require("./project/ErrorWithLocation");
Object.defineProperty(exports, "ErrorWithLocation", { enumerable: true, get: function () { return __importDefault(ErrorWithLocation_1).default; } });
var FileSystemWatchManager_1 = require("./project-api/FileSystemWatchManager");
Object.defineProperty(exports, "EventType", { enumerable: true, get: function () { return FileSystemWatchManager_1.EventType; } });
var ItemFilterImpl_1 = require("./project-api/ItemFilterImpl");
Object.defineProperty(exports, "ItemFilterImpl", { enumerable: true, get: function () { return __importDefault(ItemFilterImpl_1).default; } });
__exportStar(require("./util/Types"), exports);
var makeType_1 = require("./util/makeType");
Object.defineProperty(exports, "makeType", { enumerable: true, get: function () { return __importDefault(makeType_1).default; } });
var NPMUtil_1 = require("./util/NPMUtil");
Object.defineProperty(exports, "NPMUtil", { enumerable: true, get: function () { return __importDefault(NPMUtil_1).default; } });
__exportStar(require("./logging/LoggerManager"), exports);
var ProjectAPILogger_1 = require("./logging/ProjectAPILogger");
Object.defineProperty(exports, "ProjectAPILogger", { enumerable: true, get: function () { return __importDefault(ProjectAPILogger_1).default; } });
var ConsoleLogger_1 = require("./logging/ConsoleLogger");
Object.defineProperty(exports, "consoleLogger", { enumerable: true, get: function () { return ConsoleLogger_1.consoleLogger; } });
__exportStar(require("./logging/noop-logger"), exports);
__exportStar(require("./commons/FileSystemFactory"), exports);
var PropertyFile_1 = require("./commons/PropertyFile");
Object.defineProperty(exports, "PropertyFile", { enumerable: true, get: function () { return __importDefault(PropertyFile_1).default; } });
var Util_1 = require("./util/Util");
Object.defineProperty(exports, "Util", { enumerable: true, get: function () { return __importDefault(Util_1).default; } });
var ProjectFileSystem_1 = require("./commons/ProjectFileSystem");
Object.defineProperty(exports, "ProjectFileSystem", { enumerable: true, get: function () { return __importDefault(ProjectFileSystem_1).default; } });
var NodeFileSystem_1 = require("./commons/NodeFileSystem");
Object.defineProperty(exports, "NodeFileSystem", { enumerable: true, get: function () { return NodeFileSystem_1.NodeFileSystem; } });
var RunUI_1 = require("./cp/RunUI");
Object.defineProperty(exports, "RunUI", { enumerable: true, get: function () { return __importDefault(RunUI_1).default; } });
var NameFactory_1 = require("./cp/NameFactory");
Object.defineProperty(exports, "NameFactory", { enumerable: true, get: function () { return __importDefault(NameFactory_1).default; } });
var Service_1 = require("./cp/Service");
Object.defineProperty(exports, "Service", { enumerable: true, get: function () { return Service_1.Service; } });
var DestinationData_1 = require("./cp/DestinationData");
Object.defineProperty(exports, "DestinationData", { enumerable: true, get: function () { return __importDefault(DestinationData_1).default; } });
var CapApi_1 = require("./cp/CapApi");
Object.defineProperty(exports, "CapApi", { enumerable: true, get: function () { return __importDefault(CapApi_1).default; } });
var UIManifest_1 = require("./cp/UIManifest");
Object.defineProperty(exports, "UIManifest", { enumerable: true, get: function () { return __importDefault(UIManifest_1).default; } });
var UI5ManifestHelper_1 = require("./cp/UI5ManifestHelper");
Object.defineProperty(exports, "UI5ManifestHelper", { enumerable: true, get: function () { return __importDefault(UI5ManifestHelper_1).default; } });
var UI5RunUI_1 = require("./cp/UI5RunUI");
Object.defineProperty(exports, "UI5RunUI", { enumerable: true, get: function () { return __importDefault(UI5RunUI_1).default; } });
var MtaTemplate_1 = require("./mta-generator/MtaTemplate");
Object.defineProperty(exports, "MtaTemplate", { enumerable: true, get: function () { return __importDefault(MtaTemplate_1).default; } });
var TestUtil_1 = require("./util/TestUtil");
Object.defineProperty(exports, "TestUtil", { enumerable: true, get: function () { return __importDefault(TestUtil_1).default; } });
var StateMessage_1 = require("./commons/StateMessage");
Object.defineProperty(exports, "Info", { enumerable: true, get: function () { return StateMessage_1.Info; } });
