// @sap/artifact-management/dist/src/plugins/ui5/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InfoList = void 0;
const artifact_management_base_1 = require("@sap/artifact-management-base");
const UI5ModuleReader_1 = __importDefault(require("./readers/UI5ModuleReader"));
const UI5ModuleGenerator_1 = __importDefault(require("./generators/UI5ModuleGenerator"));
const UI5Plugin = {
    name: '@sap/ui5-project-plugin',
    getDependencies() {
        return [];
    },
    getProjectReaders() {
        return [];
    },
    // TODO: Need to incorporate RunUI!!
    getModuleReaders() {
        return [new UI5ModuleReader_1.default()];
    },
    getProjectManifestGenerators() {
        return [];
    },
    getModuleManifestGenerators() {
        return [new UI5ModuleGenerator_1.default()];
    },
    getModuleRunner() {
        return new artifact_management_base_1.UI5RunUI();
    },
};
exports.default = UI5Plugin;
var InfoList_1 = require("./InfoList");
Object.defineProperty(exports, "InfoList", { enumerable: true, get: function () { return InfoList_1.InfoList; } });
