// @sap/artifact-management/dist/src/plugins/ui5/InfoList.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InfoList = void 0;
exports.InfoList = {
    ERR_MODULE_SAP_APP_ID_UNDEFINED: {
        code: 'ui5_001',
        description: 'Cannot find sap.app.id in manifest.json, hence name of this module is undefined',
    },
    ERR_METADATA_NAME_UNDEFINED: {
        code: 'ui5_002',
        description: 'Cannot find metadata.name in ui5.yaml, hence name of this module is undefined',
    },
    ERR_SAP_APP_TITLE_UNDEFINED: {
        code: 'ui5_003',
        description: 'Cannot find sap.app.title in manifest.json or its localised value in i18n.properties',
    },
    ERR_CROSSNAVIGSTION_SEMANTIC_OBJECT_UNDEFINED: {
        code: 'ui5_004',
        description: 'Cannot find semanticObject in sap.app.crossNavigation in manifest.json, hence name and ref of the item are incorrect',
    },
    ERR_CROSSNAVIGSTION_ACTION_UNDEFINED: {
        code: 'ui5_005',
        description: 'Cannot find action in sap.app.crossNavigation in manifest.json, hence name and ref of the item are incorrect',
    },
    ERR_TARGET_ID_UNDEFINED: {
        code: 'ui5_006',
        description: 'Cannot find id in sap.ui5.routing.targets object in manifest.json, hence name is undefined and ref is incorrect',
    },
    ERR_SOURCETEMPLATE_ID_UNDEFINED: {
        code: 'ui5_007',
        description: 'Cannot find sap.app.sourceTemplate.id in manifest.json, hence sourceTemplate id is undefined',
    },
    ERR_ITEM_SAP_APP_ID_UNDEFINED: {
        code: 'ui5_008',
        description: 'Cannot find sap.app.id in manifest.json, hence ref of the item is undefined',
    },
};
