// @sap/artifact-management/dist/src/plugins/xsuaa/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InfoList = void 0;
const XsSecurityModuleManifestGenerator_1 = __importDefault(require("./generators/XsSecurityModuleManifestGenerator"));
const XsSecurityModuleReader_1 = __importDefault(require("./readers/XsSecurityModuleReader"));
const XsSecurityPlugin = {
    name: '@sap/xsuaa-project-plugin',
    getDependencies() {
        return [];
    },
    getProjectReaders() {
        return [];
    },
    getModuleReaders() {
        return [new XsSecurityModuleReader_1.default()];
    },
    getProjectManifestGenerators() {
        return [];
    },
    getModuleManifestGenerators() {
        return [new XsSecurityModuleManifestGenerator_1.default()];
    },
};
exports.default = XsSecurityPlugin;
var InfoList_1 = require("./InfoList");
Object.defineProperty(exports, "InfoList", { enumerable: true, get: function () { return InfoList_1.InfoList; } });
