// @sap/artifact-management/dist/src/mta-generator/MtaBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class MtaBuilder {
    constructor(fs, mta) {
        this.beforeBuildCommands = [];
        this.modules = [];
        this.resources = [];
        this.fs = fs;
        this.mta = mta;
    }
    addModule(...modules) {
        this.modules = this.modules.concat(modules);
    }
    getModules() {
        return this.modules;
    }
    addResource(...resources) {
        this.resources = this.resources.concat(resources);
    }
    getResources() {
        return this.resources;
    }
    addBeforeBuildCommand(...command) {
        this.beforeBuildCommands.push(...command);
    }
    async generate() {
        const mta = Object.assign({}, this.mta);
        mta.resources = (mta.resources || []).concat(this.resources);
        mta.modules = (mta.modules || []).concat(this.modules);
        if (this.beforeBuildCommands.length > 0) {
            mta['build-parameters'] = mta['build-parameters'] || {};
            mta['build-parameters']['before-all'] = mta['build-parameters']['before-all'] || [];
            mta['build-parameters']['before-all'] = [{
                    builder: 'custom',
                    'commands': this.beforeBuildCommands,
                }].concat(mta['build-parameters']['before-all']);
        }
        return mta;
    }
}
exports.default = MtaBuilder;
